From 954f1e1c7c3bf63a4c61d880a5dc8d43b2dd5a8e Mon Sep 17 00:00:00 2001
From: Andy Ross <andy.ross@windriver.com>
Date: Mon, 28 Jan 2013 16:20:04 -0800
Subject: init: add DMI-detected machine properties.

At boot time, check for DMI and match vs. a
/system/etc/dmi-machine.conf file to load a property file from
/system/etc/machine-props.

For: AXIA-1751
Change-Id: Ifdda600bff44b0b39cf2ac6d0b0211fc2c36b977
Signed-off-by: Andy Ross <andy.ross@windriver.com>
---
 init/property_service.c |   50 +++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 50 insertions(+), 0 deletions(-)

diff --git a/init/property_service.c b/init/property_service.c
index eb34d71..c2efae9 100755
--- a/init/property_service.c
+++ b/init/property_service.c
@@ -579,6 +579,55 @@ static void load_persistent_properties()
     persistent_properties_loaded = 1;
 }
 
+static void dmi_detect_props(void)
+{
+    char dmi[256], buf[256], path[256], *hash, *tag, *name, *toksav, allmatched;
+    const char *var;
+    FILE *dmif, *cfg;
+
+    /* Skip if disabled */
+    var = property_get("ro.boot.dmi_detect");
+    if(var && strcmp(var, "0"))
+        return;
+
+    /* Fail silently (no error) on devices without DMI */
+    dmif = fopen("/sys/devices/virtual/dmi/id/modalias", "r");
+    if (!dmif || !fgets(dmi, sizeof(dmi), dmif))
+        return;
+    fclose(dmif);
+
+    /* ...or if the system filesystem lacks configuration */
+    if (!(cfg = fopen("/system/etc/dmi-machine.conf", "r")))
+        return;
+
+    while (fgets(buf, sizeof(buf), cfg)) {
+        /* snip comments */
+        if ((hash = strchr(buf, '#')))
+            *hash = 0;
+
+        /* parse line: cfg name, then a list of tags to test.  If they
+         * all match then load the config file and return (note that
+         * "no tags" means "everything matched" and can be used to
+         * load a default if needed, though that is probably best done
+         * with build.props) */
+        allmatched = 1;
+        if ((name = strtok_r(buf, " \t\v\r\n", &toksav))) {
+            while ((tag = strtok_r(NULL, " \t\v\r\n", &toksav)))
+                if (!strstr(dmi, tag))
+                    allmatched = 0;
+
+            if (allmatched) {
+                property_set("ro.sys.detected_machine", name);
+                snprintf(path, sizeof(path), "/system/etc/machine-props/%s.prop", name);
+                load_properties_from_file(path);
+                break;
+            }
+        }
+    }
+
+    fclose(cfg);
+}
+
 void property_init(void)
 {
     init_property_area();
@@ -620,6 +669,7 @@ void start_property_service(void)
 {
     int fd;
 
+    dmi_detect_props();
     load_properties_from_file(PROP_PATH_SYSTEM_BUILD);
     load_properties_from_file(PROP_PATH_SYSTEM_DEFAULT);
     load_override_properties();
-- 
1.7.4.1

